### PR TITLE
Fix repeated cipher detection

### PR DESCRIPTION
**Pull request by jsherwood0**
In our assessments we ran into servers that generated infinite lists of the same cipher, causing sslscan to loop forever.  By changing markFoundCipherSuite to return false if that cipher suite was already found, and then branching to done if that happens, we eliminated that issue.

### CODE REVIEW

**Comment by rbsec** [comment]
That's an interesting bug, and not one I've run into before. Do you know what the software/versions on the server was?

Thanks for submitting a fix (rather than just an issue). It looks good to me - @jtesta do you have any comments?

**Comment by jsherwood0** [comment]
The bulk of the servers we saw this on were Citrix web servers (NetScaler, Access Gateway, Receiver, and Director).  No idea what versions though since the pages actively avoid providing version information.  Likely pretty recent though.  The Citrix Receiver executable version downloaded from the Citrix Receiver web server showed as v14.9.6000.6.

**Comment by jtesta** [comment]
I actually ran into this issue myself a couple weeks ago, but I got side-tracked and didn't finish investigating.

When we look for cipher suites not implemented by OpenSSL, apparently some servers will break the TLS spec and respond with a cipher suite ID that we didn't request (!).  This was confusing the code and causing an infinite loop.

I took a look at this PR.  I think a side-effect to the way the fix is implemented is that it will report the unrequested cipher ID twice.  I implemented my own fix in PR #206 that handles this issue a little more cleanly.

@jsherwood0: can you please test my fix and see if it works for you as well?

**Comment by jsherwood0** [comment]
It looks like your fix handles the instance that we were seeing too.  And just to make sure I understand, you iterate through your list of ciphers, and if it replies with one not matching what you asked for, you consider it bogus and move on, correct?  If so, is there some way to indicate the non-standard/unexpected cipher ID in the output?  The reason I'm asking is that in my test case the old code shows the following over and over, but the new code doesn't show it at all:

Accepted  TLSv1.2  256 bits  TLS_RSA_WITH_AES_256_CBC_SHA

It's unclear to me whether this is a cipher that the server is trying to force (in which case failing to list it would be the wrong thing to do), or it is an artifact of a junk response that happens to line up with that cipher in your struct (in which case it really is junk to be omitted).

**Comment by jtesta** [comment]
> It looks like your fix handles the instance that we were seeing too. And 
> just to make sure I understand, you iterate through your list of 
> ciphers, and if it replies with one not matching what you asked for, you 
> consider it bogus and move on, correct?

More than just that, actually: it stops doing the missing cipher tests 
entirely.  Any further testing would also produce invalid IDs, so 
there's no point in continuing.

It seems like the server has no matching cipher suites, so its simply 
responding back with one it does have.  So that's how sslscan interprets 
this (broken) behavior.


> If so, is there some way to 
> indicate the non-standard/unexpected cipher ID in the output?

It would be nice to have a debug output option, but there isn't one at 
present.


> The reason 
> I'm asking is that in my test case the old code shows the following over 
> and over, but the new code doesn't show it at all:
> 
> Accepted TLSv1.2 256 bits TLS_RSA_WITH_AES_256_CBC_SHA

There's a convoluted reason for this.

The issue is that IANA assigned official names to the ciphers, but 
OpenSSL decided (for some reason) to invent their own names.  The bulk 
of the ciphersuite testing in sslscan is done with OpenSSL, hence it 
outputs the OpenSSL names.  Months ago, I found that OpenSSL doesn't 
implement all RFC-defined ciphersuites, so we added a fallback mechanism 
to test all missing ciphersuites (this was the part that was spinning in 
the infinite loop).  Since these ciphersuites aren't implemented in 
OpenSSL, they don't have OpenSSL names, so their IANA names must be used.

Without my fix, the missing-ciphersuites code was getting back 
unrequested cipher ID 0x0035, which IANA calls 
"TLS_RSA_WITH_AES_256_CBC_SHA" and OpenSSL calls "AES256-SHA".  So you 
were seeing "AES256-SHA" being reported properly once, then the infinite 
loop reporting "TLS_RSA_WITH_AES_256_CBC_SHA" over and over again 
(because the missing ciphersuites code wasn't prepared for this broken 
server behavior).

With my fix, cipher ID 0x0035 will be properly found by OpenSSL and 
reported once as "AES256-SHA".  Then the missing ciphersuites fallback 
will notice an invalid cipher ID and conclude there are no more 
ciphersuites supported by the server.  Hence its IANA name of 
"TLS_RSA_WITH_AES_256_CBC_SHA" is never reported (which is correct, 
since it was already reported under its OpenSSL name).

Long story short... I believe sslscan is now giving proper output. 
Exactly all the ciphersuites supported by the server are being reported 
correctly.

-- 
Joseph S. Testa II
Founder & Principal Security Consultant
Positron Security

**Comment by jsherwood0** [comment]
Sounds good then.  I appreciate the quick turnaround on this comprehensive fix.  Normally I don't like submitting an issue report without a proposed fix, but your fix is definitely the better choice based on the additional info that you provided.  I'm going to close this PR in favor of yours.

When do you expect your fix to be merged into the master branch?  Once it's merged I'll have our systems run with the git build until Kali catches up.

Regards,

-John Sherwood

**Comment by rbsec** [comment]
That is an interesting bug in their TLS stack - I can imagine it's one that's not really been discovered before. If you have enough information it might be worth reporting it to them (although I have no idea what Citrix are like for that, so it might be a painful process). It's amazing how buggy most TLS implementations are - there are a whole load of legacy workaround in the original sslcsan code to try and accommodate some of them.

Thanks for the details bug report, and also to @jtesta for putting together such a quick fix. I've merged #206 into master, so I'll close this is out in favour of the fix from @jtesta, and tag a new version.
